### PR TITLE
Reorganize DTF pattern and move it to reference module.

### PR DESCRIPTION
**Pull request by zbraniecki**
This PR is doing some move-around, mostly moving the `Pattern` to `reference::Pattern` and making space for `runtime::Pattern` backed by ZeroVec.

The rest is just clean up or things that accrued around.

Two things I *did* not clean up is parser and generic pattern. I'll do it after I introduce runtime pattern and move provider to it.

### CODE REVIEW

**Comment by zbraniecki** [comment]
Part of #519

**Review by dminor — APPROVED** [review]
lgtm

File: components/datetime/src/pattern/reference/parser.rs [hunk 1]
```diff
@@ -235,19 +237,22 @@ impl<'p> Parser<'p> {
...
-    pub fn parse_placeholders(self, replacements: Vec<Pattern>) -> Result<Vec<PatternItem>, Error> {
+    pub fn parse_placeholders(
+        self,
+        replacements: Vec<Pattern>,
+    ) -> Result<Vec<PatternItem>, PatternError> {
         let generic_items = self.parse_generic()?;
 
         let gp = GenericPattern::from(generic_items);
-        Ok(gp.combined(replacements)?.items)
+        Ok(gp.combined(replacements)?.items.to_vec())
```
Comment (gregtatum) [inline components/datetime/src/pattern/reference/parser.rs:247]: Question: Why is this `to_vec` now?

File: components/datetime/src/pattern/mod.rs [hunk 2]
```diff
@@ -71,264 +19,45 @@ impl From<char> for PatternItem {
...
-                }
-            }
-        }
-        dump_buffer_into_formatter(&buffer, formatter)?;
-        buffer.clear();
-        Ok(())
-    }
-}
-
-impl FromIterator<PatternItem> for Pattern {
```
Comment (gregtatum) [inline components/datetime/src/pattern/mod.rs:223]: Question: I guess this is going away?

File: components/datetime/src/pattern/reference/pattern.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+#[cfg(feature = "provider_serde")]
+use serde::{
+    de,
+    ser::{self, SerializeSeq},
+    Deserialize, Deserializer, Serialize,
+};
+
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct Pattern {
+    pub items: Vec<PatternItem>,
```
[the remark this answers is not in the document]
Comment (gregtatum) [reply components/datetime/src/pattern/reference/pattern.rs:26]: Actually looking below the items are already accessible below. I don't think we should have these the `items()` and `items_mut` as well as `pub` fields. I don't think there is any benefit here.

It appears you could already publicly access the pattern, even if it isn't publicly documented.

**Review by gregtatum — COMMENTED** [review]
This seems to be merged while I was doing a review. I believe this merge was against the review policy we have outlined here: https://github.com/unicode-org/icu4x/blob/a0d4cf41951316c7f1c8733e20e8577fcd7e3f51/CONTRIBUTING.md#pr-author-and-reviewers-workflow

I had a bit of feedback and questions that would have been nice to answer and address before merging.

One other piece of feedback I would call out is that this review took quite awhile to do because it was combining code moves and code changes, which took quite awhile to read through and separate out. I know ICU4X doesn't have a culture of clean commit history, but this would have been pretty trivial to review most of it if the commits were broken out a bit better.

File: components/datetime/src/pattern/reference/mod.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
```
[the remark this answers is not in the document]
Comment (zbraniecki) [reply components/datetime/src/pattern/reference/mod.rs:4]: I'll add documentation after I'm done with refactors!

File: components/datetime/src/pattern/mod.rs [hunk 5]
```diff
@@ -71,264 +19,45 @@ impl From<char> for PatternItem {
...
-    fn visit_seq<V>(self, mut seq: V) -> Result<Pattern, V::Error>
-    where
-        V: de::SeqAccess<'de>,
-    {
-        let mut items = Vec::new();
-        while let Some(item) = seq.next_element()? {
-            items.push(item)
-        }
-        Ok(Pattern::from(items))
+impl Default for TimeGranularity {
```
[the remark this answers is not in the document]
Comment (zbraniecki) [reply components/datetime/src/pattern/mod.rs:29]: I didn't think of it as a code change, but I recognize that it is now. I apologize for introducing it here.

File: components/datetime/src/pattern/reference/pattern.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zbraniecki) [reply components/datetime/src/pattern/reference/pattern.rs:26]: I'm shifting between two types of iterators on items, and I haven't figure out how the mutability of items will work on runtime Pattern yet. It's a mid-step. I'll clean up pub/non-pub before release.

File: components/datetime/src/fields/length.rs [hunk 6]
```diff
@@ -31,38 +28,24 @@ pub enum FieldLength {
...
-        match length {
-            FieldLength::One => 1,
-            FieldLength::TwoDigit => 2,
-            FieldLength::Abbreviated => 3,
-            FieldLength::Wide => 4,
-            FieldLength::Narrow => 5,
-            FieldLength::Six => 6,
-        }
+impl FieldLength {
+    pub fn u8_in_range(v: &u8) -> bool {
```
[the remark this answers is not in the document]
Comment (zbraniecki) [reply components/datetime/src/fields/length.rs:32]: This is going to be used by ULE.

[the remark this answers is not in the document]
Comment (zbraniecki) [reply components/datetime/src/fields/length.rs:32]: Good point. I'll unify that - I originally wrote it to use just `u8` and even had From/Into, but I realized that the `u8` itself can mean different things so for operations between `u8` as an index of a variant of an Enum to encode/decode I should use `from_idx` and `idx`. I'll update it while cleaning up!

File: components/datetime/src/pattern/reference/pattern.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        Parser::new(input)
+            .parse_placeholders(vec![time, date])
+            .map(Self::from)
+    }
+}
+
+impl From<Vec<PatternItem>> for Pattern {
+    fn from(items: Vec<PatternItem>) -> Self {
+        Self {
+            time_granularity: items.iter().map(Into::into).max().unwrap_or_default(),
```
[the remark this answers is not in the document]
Comment (zbraniecki) [reply components/datetime/src/pattern/reference/pattern.rs:58]: That's how I always saw this code written, but I'll clean it up after I finish the move to make it more readable!

File: components/datetime/src/pattern/mod.rs (hunk 2 above)
Comment (zbraniecki) [reply components/datetime/src/pattern/mod.rs:223]: Haven't been used anywhere, I removed unused traits.

File: components/datetime/src/fields/length.rs [hunk 8]
```diff
@@ -31,38 +28,24 @@ pub enum FieldLength {
...
+        (1..=6).contains(v)
     }
-}
 
-macro_rules! try_field_length {
-    ($i:ty) => {
-        impl TryFrom<$i> for FieldLength {
-            type Error = LengthError;
+    pub fn idx(&self) -> u8 {
```
[the remark this answers is not in the document]
Comment (sffc) [reply components/datetime/src/fields/length.rs:36]: +1; in general, all exported symbols should have documentation.  I don't care if it's internal, but I'm pretty sure `FieldLength` is exported, which means `FieldLength::idx` is exported.

**Review by sffc — COMMENTED** [review]
(post-submit review)

Looks like mostly a fairly straightforward refactor.  I did not double-check the added code to make sure that it is 100% equal to the original code.  As Greg said, my main feedback would be to use this opportunity to make sure that all exported symbols have docstrings.

**Pull request merged**